src/main/java/teammates/common/util/ActivityLogGenerator.java

```diff
@@ -0,0 +1,336 @@
+                       + builder.getActionServletName() + "<br>"
+                       + e.getClass() + ": " + TeammatesException.toStringWithStackTrace(e) + "<br>"
+                       + JsonUtils.toJson(params, Map.class) + "</span>";
+        builder.withLogMessage(message);
+        
+        builder.withActionName(Const.ACTION_RESULT_FAILURE);
+        
+        return builder.build().generateLogMessage();
+    }
+    
+    /**
+     * Generates log message for system error report.
+     * <p>
+     * If the user has logged in with google id or authenticated key, his/her identity will be recorded
+     * in the log id.
+     * 
+     * @param url URL of the request
+     * @param params parameterMap of the request
+     * @param errorEmail EmailWrapper to send to the system admin
+     * @param currUser login information generated by {@link teammates.logic.api.GateKeeper}
+     * @return log message in form specified in {@link ActivityLogEntry}
+     */
+    public String generateSystemErrorReportLogMessage(String url, Map<String, String[]> params,
+                                                      EmailWrapper errorEmail, UserType currUser) {
+        ActivityLogEntry.Builder builder = generateBasicActivityLogEntry(url, params, currUser);
+        
+        // TODO can use template to generate this
+        if (errorEmail != null) {
+            String message = "<span class=\"text-danger\">" + errorEmail.getSubject() + "</span>"
+                    + "<br>"
+                    + "<a href=\"#\" onclick=\"showHideErrorMessage('error" + errorEmail.hashCode() + "');\">"
+                        + "Show/Hide Details >>"
+                    + "</a>"
+                    + "<br>"
+                    + "<span id=\"error" + errorEmail.hashCode() + "\" style=\"display: none;\">"
+                        + errorEmail.getContent()
+                    + "</span>";
+            builder.withLogMessage(message);
+        }
+        
+        builder.withActionName(Const.ACTION_RESULT_SYSTEM_ERROR_REPORT);
+       
+        return builder.build().generateLogMessage();
+    }
+    
+    /**
+     * Generates basic builder for activityLogEntry.
+     * 
+     * @param url URL of the request
+     * @param params parameterMap of the request
+     * @param req HttpServletRequest in the request
+     * @param currUser login information generated by {@link teammates.logic.api.GateKeeper}
+     * @return ActivityLogEntry.Builder builder with basic information
+     */
+    private ActivityLogEntry.Builder generateBasicActivityLogEntry(String url, Map<String, String[]> params,
+                                                                   UserType currUser) {
+        String servletName = getActionNameFromUrl(url);
+        long currTime = System.currentTimeMillis();
+        ActivityLogEntry.Builder builder = new ActivityLogEntry.Builder(servletName, url, currTime);
+        builder.withActionName(servletName); // by default
```

Comment:
Hmm.. What do you mean by "by default"? Why call this method when the field is set by the constructor too?